helix-vcs/src/rope.rs

```diff
@@ -0,0 +1,20 @@
+use std::hash::Hash;
+
+use ropey::Rope;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct RopeLine<'a>(pub ropey::RopeSlice<'a>);
+
+impl Hash for RopeLine<'_> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        for chunk in self.0.chunks() {
+            chunk.hash(state);
+        }
+    }
+}
+
+impl<'a> RopeLine<'a> {
+    pub fn from_rope(rope: &'a Rope) -> Vec<Self> {
+        rope.lines().into_iter().map(RopeLine).collect()
```

Comment:
Note: `.lines()` is not very efficient: https://github.com/cessen/ropey/issues/25

Reply:
Ah I read `O(1)` in the chunks iterator and assumed lines would also follow suit. @cessen how hard/big of a change would it be to make the lines iterator also O(1) ? I suppose if it was easy enough it would've already been added.

Reply:
From casual use it seems snappy and I haven't noticed any performance issues, so this is probably not too serious of an issue :)

Reply:
> But then again this is a hotpath since it's called on almost every keystroke

Synchronously?

Thinking about it more, I would expect the diffing itself to be expensive enough that if the `Lines` iterator is causing performance issues then the diffing would be as well (i.e. even if `Lines` was infinitely fast, you'd still have issues).  So if this feature does turn out to cause performance problems (e.g. on large files or with large diffs), then rearchitecting it to work in a separate thread or something like that might be needed anyway.